File: src/main/java/com/bajdcc/LALR1/interpret/module/api/ModuleNetWebApiHelper.java
package com.bajdcc.LALR1.interpret.module.api;

import com.bajdcc.LALR1.grammar.runtime.RuntimeObject;
import com.bajdcc.LALR1.grammar.runtime.RuntimeObjectType;
import com.bajdcc.LALR1.grammar.runtime.data.RuntimeArray;
import com.bajdcc.LALR1.grammar.runtime.data.RuntimeMap;

import java.util.Map;
import java.util.stream.Collectors;

/**
 * 实用方法
 *
 * @author bajdcc
 */
public class ModuleNetWebApiHelper {

	public static Object toJsonObject(RuntimeObject obj) {
		if (obj == null)
			return null;
		switch (obj.getType()) {
			case kNull:
			case kFunc:
			case kNan:
				return null;
			case kPtr:
			case kObject:
			case kChar:
			case kString:
			case kBool:
			case kInt:
			case kReal:
				return obj.getObj();
			case kArray:
				return ((RuntimeArray) obj.getObj()).getArray().stream()
						.map(ModuleNetWebApiHelper::toJsonObject).collect(Collectors.toList());
			case kMap:
				return ((RuntimeMap) obj.getObj()).getMap().entrySet().stream()
						.filter(entry -> entry.getValue().getType() != RuntimeObjectType.kNull)
						.collect(Collectors.toMap(Map.Entry::getKey, entry -> toJsonObject(entry.getValue())));
		}
		return null;
	}
}

File: src/main/kotlin/com/bajdcc/LALR1/interpret/module/api/ModuleNetWebApiHelper.kt
package com.bajdcc.LALR1.interpret.module.api

import com.bajdcc.LALR1.grammar.runtime.RuntimeObject
import com.bajdcc.LALR1.grammar.runtime.RuntimeObjectType
import com.bajdcc.LALR1.grammar.runtime.data.RuntimeArray
import com.bajdcc.LALR1.grammar.runtime.data.RuntimeMap

/**
 * 实用方法
 *
 * @author bajdcc
 */
object ModuleNetWebApiHelper {

    fun toJsonObject(obj: RuntimeObject?): Any? {
        if (obj == null)
            return null
        when (obj.type) {
            RuntimeObjectType.kNull, RuntimeObjectType.kFunc, RuntimeObjectType.kNan -> return null
            RuntimeObjectType.kPtr, RuntimeObjectType.kObject, RuntimeObjectType.kChar, RuntimeObjectType.kString, RuntimeObjectType.kBool, RuntimeObjectType.kInt, RuntimeObjectType.kReal -> return obj.obj
            RuntimeObjectType.kArray -> return (obj.obj as RuntimeArray).getArray()
                    .map { toJsonObject(it) }
            RuntimeObjectType.kMap -> return (obj.obj as RuntimeMap).getMap().entries
                    .filter { it.value.type !== RuntimeObjectType.kNull }
                    .associate { it.key to toJsonObject(it.value) }
            else -> {
            }
        }
        return null
    }
}
